bulkcopy.go

```diff
@@ -0,0 +1,576 @@
+
+	case typeBit, typeBitN:
+		if reflect.TypeOf(val).Kind() != reflect.Bool {
+			err = fmt.Errorf("mssql: invalid type for bit column", val)
+			return
+		}
+		res.ti.TypeId = typeBitN
+		res.ti.Size = 1
+		res.buffer = make([]byte, 1)
+		if val.(bool) {
+			res.buffer[0] = 1
+		}
+
+	case typeDateTime2N, typeDateTimeOffsetN:
+		switch val := val.(type) {
+		case time.Time:
+			days, ns := dateTime2(val)
+			ns /= int64(math.Pow10(int(col.ti.Scale)*-1) * 1000000000)
+
+			var data = make([]byte, 5)
+
+			data[0] = byte(ns)
+			data[1] = byte(ns >> 8)
+			data[2] = byte(ns >> 16)
+			data[3] = byte(ns >> 24)
+			data[4] = byte(ns >> 32)
+
+			if col.ti.Scale <= 2 {
+				res.ti.Size = 6
+			} else if col.ti.Scale <= 4 {
+				res.ti.Size = 7
+			} else {
+				res.ti.Size = 8
+			}
+			var buf []byte
+			buf = make([]byte, res.ti.Size)
+			copy(buf, data[0:res.ti.Size-3])
+
+			buf[res.ti.Size-3] = byte(days)
+			buf[res.ti.Size-2] = byte(days >> 8)
+			buf[res.ti.Size-1] = byte(days >> 16)
+
+			if col.ti.TypeId == typeDateTimeOffsetN {
+				_, offset := val.Zone()
+				var offsetMinute = uint16(offset / 60)
+				buf = append(buf, byte(offsetMinute))
+				buf = append(buf, byte(offsetMinute>>8))
+				res.ti.Size = res.ti.Size + 2
+			}
+
+			res.buffer = buf
+
+		default:
+			err = fmt.Errorf("mssql: invalid type for datetime2 column", val)
+			return
+		}
+	case typeDateN:
+		switch val := val.(type) {
+		case time.Time:
+			days, _ := dateTime2(val)
+
+			res.ti.Size = 3
+			res.buffer = make([]byte, 3)
+			res.buffer[0] = byte(days)
+			res.buffer[1] = byte(days >> 8)
+			res.buffer[2] = byte(days >> 16)
+		default:
+			err = fmt.Errorf("mssql: invalid type for date column", val)
+			return
+		}
+	case typeDateTime, typeDateTimeN, typeDateTim4:
+		switch val := val.(type) {
+		case time.Time:
+			if col.ti.Size == 4 {
+				res.ti.Size = 4
+				res.buffer = make([]byte, 4)
+
+				ref := time.Date(1900, 1, 1, 0, 0, 0, 0, time.UTC)
+				dur := val.Sub(ref)
+				days := dur / (24 * time.Hour)
+				if days < 0 {
+					err = fmt.Errorf("mssql: Date %s is out of range", val)
+					return
+				}
+				mins := val.Hour()*60 + val.Minute()
+
+				binary.LittleEndian.PutUint16(res.buffer[0:2], uint16(days))
+				binary.LittleEndian.PutUint16(res.buffer[2:4], uint16(mins))
+			} else if col.ti.Size == 8 {
+				res.ti.Size = 8
+				res.buffer = make([]byte, 8)
+
+				ref := time.Date(1900, 1, 1, 0, 0, 0, 0, time.UTC)
+				dur := val.Sub(ref)
+				days := math.Floor(float64(dur) / float64(24*time.Hour))
+				tm := (val.Hour()*60*60+val.Minute()*60+val.Second())*300 + int(val.Nanosecond()/10000000*3)
+
+				binary.LittleEndian.PutUint32(res.buffer[0:4], uint32(days))
+				binary.LittleEndian.PutUint32(res.buffer[4:8], uint32(tm))
+			} else {
+				err = fmt.Errorf("mssql: invalid size of column")
+			}
+
+		default:
+			err = fmt.Errorf("mssql: invalid type for datetime column", val)
+		}
+
+	/*
+		case typeDecimal, typeDecimalN:
```

Comment:
I'm fine to not support decimal and numeric at this time. We can address later. Feel free to leave commented out.